File: packages/utils/src/value-type-and-unit.ts
/** What an indicator's values are, in the form's order. Each id is its `value_type` row's. */
export const VALUE_TYPES = {
  count: { id: '01a0d8a5-3ca2-7315-bfca-96c7324d4347', name: 'Count' },
  crudeRate: { id: '01a0d8a5-3ca2-7315-bfca-96c8c77cad18', name: 'Crude rate' },
  directlyStandardisedRate: {
    id: '01a0d8a5-3ca2-7315-bfca-96c9664c846c',
    name: 'Directly standardised rate',
  },
  excessRisk: { id: '01a0d8a5-3ca2-7315-bfca-96ca4c9608e9', name: 'Excess risk' },
  gap: { id: '01a0d8a5-3ca2-7315-bfca-96cb03846bff', name: 'Gap' },
  indirectlyStandardisedProportion: {
    id: '01a0d8a5-3ca2-7315-bfca-96cc2dc711c3',
    name: 'Indirectly standardised proportion',
  },
  indirectlyStandardisedRatio: {
    id: '01a0d8a5-3ca2-7315-bfca-96cda63ea940',
    name: 'Indirectly standardised ratio',
  },
  lifeExpectancy: { id: '01a0d8a5-3ca2-7315-bfca-96ce5ffc8d47', name: 'Life expectancy' },
  mean: { id: '01a0d8a5-3ca2-7315-bfca-96cff286704d', name: 'Mean' },
  median: { id: '01a0d8a5-3ca2-7315-bfca-96d0928fd186', name: 'Median' },
  percentagePoint: { id: '01a0d8a5-3ca2-7315-bfca-96d1b508d83f', name: 'Percentage point' },
  proportion: { id: '01a0d8a5-3ca2-7315-bfca-96d2031a65e7', name: 'Proportion' },
  ratio: { id: '01a0d8a5-3ca2-7315-bfca-96d34a17c74b', name: 'Ratio' },
  score: { id: '01a0d8a5-3ca2-7315-bfca-96d4cc7f7bcb', name: 'Score' },
  slopeIndexOfInequality: {
    id: '01a0d8a5-3ca2-7315-bfca-96d54168d21f',
    name: 'Slope index of inequality',
  },
} as const;

/** The units of an indicator's values, in the form's order. Each id is its `unit` row's. */
export const UNITS = {
  percent: { id: '01a0d8a5-3ca2-7315-bfca-96d615820bd4', name: '%' },
  per100: { id: '01a0d8a5-3ca2-7315-bfca-96d787920e40', name: 'per 100' },
  per1000: { id: '01a0d8a5-3ca2-7315-bfca-96d8d8d7aec2', name: 'per 1,000' },
  per10000: { id: '01a0d8a5-3ca2-7315-bfca-96d97b88dbff', name: 'per 10,000' },
  per100000: { id: '01a0d8a5-3ca2-7315-bfca-96daa0c93ee9', name: 'per 100,000' },
  per1000000: { id: '01a0d8a5-3ca2-7315-bfca-96dbaf432da2', name: 'per 1,000,000' },
  minutes: { id: '01a0d8a5-3ca2-7315-bfca-96dca2ab3487', name: 'minutes' },
  hours: { id: '01a0d8a5-3ca2-7315-bfca-96dd9ae52dc5', name: 'hours' },
  days: { id: '01a0d8a5-3ca2-7315-bfca-96de5fe00871', name: 'days' },
  weeks: { id: '01a0d8a5-3ca2-7315-bfca-96df3a570e5d', name: 'weeks' },
  months: { id: '01a0d8a5-3ca2-7315-bfca-96e0128fd88d', name: 'months' },
  years: { id: '01a0d8a5-3ca2-7315-bfca-96e1a3b040d3', name: 'years' },
  pounds: { id: '01a0d8a5-3ca2-7315-bfca-96e2d3cf5df0', name: '£' },
  poundsPerCapita: { id: '01a0d8a5-3ca2-7315-bfca-96e3aae4dc98', name: '£ per capita' },
  noUnit: { id: '01a0d8a5-3ca2-7315-bfca-96e4a96bfc8e', name: 'No unit' },
  // Named by the publisher in `unit_other`.
  other: { id: '01a0d8a5-3ca2-7315-bfca-96e5cee57158', name: 'Other' },
} as const;

export const UNIT_OTHER_MAX_LENGTH = 100;

/** The value types standardised against a reference population the publisher names. */
export const INDIRECTLY_STANDARDISED_VALUE_TYPE_IDS: readonly string[] = [
  VALUE_TYPES.indirectlyStandardisedProportion.id,
  VALUE_TYPES.indirectlyStandardisedRatio.id,
];

/** How a value type is standardised, which decides what the form asks of its population. */
export function standardisationOf(valueTypeId: string): 'direct' | 'indirect' | null {
  if (valueTypeId === VALUE_TYPES.directlyStandardisedRate.id) return 'direct';
  return INDIRECTLY_STANDARDISED_VALUE_TYPE_IDS.includes(valueTypeId) ? 'indirect' : null;
}

/** The standard population of a directly standardised rate, or one the publisher names. */
export const STANDARD_POPULATIONS = ['esp-2013', 'other'] as const;

export type StandardPopulation = (typeof STANDARD_POPULATIONS)[number];

export const STANDARD_POPULATION_LABELS: Readonly<Record<StandardPopulation, string>> = {
  'esp-2013': '2013 European Standard Population',
  other: 'Other',
};

/** The unit as the public reads it beside a value: null when the values have none. */
export function unitLabel(unit: { id: string; name: string }, other: string | null): string | null {
  if (unit.id === UNITS.noUnit.id) return null;
  return unit.id === UNITS.other.id && other !== null ? other : unit.name;
}

File: packages/db/data/seed/export/value_type_and_unit.py
"""The service's value type and unit for each Fingertips value type and unit name.

The service holds them as references to the value_type and unit rows whose ids are in
@fphd/utils/value-type-and-unit, with the unit's own name in unit_other when it is not one
of the service's; the exports translate Pholio's lookup rows to them. Migration 0026 holds
the same translation for databases that already had the Fingertips rows.
"""

VALUE_TYPE_IDS = {
    "Count": "01a0d8a5-3ca2-7315-bfca-96c7324d4347",
    "Crude rate": "01a0d8a5-3ca2-7315-bfca-96c8c77cad18",
    "Directly standardised rate": "01a0d8a5-3ca2-7315-bfca-96c9664c846c",
    "Excess risk": "01a0d8a5-3ca2-7315-bfca-96ca4c9608e9",
    "Gap": "01a0d8a5-3ca2-7315-bfca-96cb03846bff",
    "Indirectly standardised proportion": "01a0d8a5-3ca2-7315-bfca-96cc2dc711c3",
    "Indirectly standardised ratio": "01a0d8a5-3ca2-7315-bfca-96cda63ea940",
    "Life expectancy": "01a0d8a5-3ca2-7315-bfca-96ce5ffc8d47",
    "Mean": "01a0d8a5-3ca2-7315-bfca-96cff286704d",
    "Median": "01a0d8a5-3ca2-7315-bfca-96d0928fd186",
    "Percentage point": "01a0d8a5-3ca2-7315-bfca-96d1b508d83f",
    "Proportion": "01a0d8a5-3ca2-7315-bfca-96d2031a65e7",
    "Ratio": "01a0d8a5-3ca2-7315-bfca-96d34a17c74b",
    "Score": "01a0d8a5-3ca2-7315-bfca-96d4cc7f7bcb",
    "Slope index of inequality": "01a0d8a5-3ca2-7315-bfca-96d54168d21f",
}

UNIT_IDS = {
    "%": "01a0d8a5-3ca2-7315-bfca-96d615820bd4",
    "per 100": "01a0d8a5-3ca2-7315-bfca-96d787920e40",
    "per 1,000": "01a0d8a5-3ca2-7315-bfca-96d8d8d7aec2",
    "per 10,000": "01a0d8a5-3ca2-7315-bfca-96d97b88dbff",
    "per 100,000": "01a0d8a5-3ca2-7315-bfca-96daa0c93ee9",
    "per 1,000,000": "01a0d8a5-3ca2-7315-bfca-96dbaf432da2",
    "minutes": "01a0d8a5-3ca2-7315-bfca-96dca2ab3487",
    "hours": "01a0d8a5-3ca2-7315-bfca-96dd9ae52dc5",
    "days": "01a0d8a5-3ca2-7315-bfca-96de5fe00871",
    "weeks": "01a0d8a5-3ca2-7315-bfca-96df3a570e5d",
    "months": "01a0d8a5-3ca2-7315-bfca-96e0128fd88d",
    "years": "01a0d8a5-3ca2-7315-bfca-96e1a3b040d3",
    "£": "01a0d8a5-3ca2-7315-bfca-96e2d3cf5df0",
    "£ per capita": "01a0d8a5-3ca2-7315-bfca-96e3aae4dc98",
    "No unit": "01a0d8a5-3ca2-7315-bfca-96e4a96bfc8e",
    "Other": "01a0d8a5-3ca2-7315-bfca-96e5cee57158",
}

# Fingertips value types with no service value, such as Months life lost, stop the export.
VALUE_TYPES = {
    **{name: name for name in VALUE_TYPE_IDS},
    "Slope Index of Inequality": "Slope index of inequality",
    "Number": "Count",
    "Rate ratio": "Ratio",
}

# Any other named Fingertips unit is an other unit; a placeholder has no name to keep.
UNITS = {
    "Percent": "%",
    "per 100": "per 100",
    "per 1,000": "per 1,000",
    "per 10,000": "per 10,000",
    "per 100,000": "per 100,000",
    "per 1,000,000": "per 1,000,000",
    "Minutes": "minutes",
    "Days": "days",
    "Years": "years",
    "£": "£",
    "£ per capita": "£ per capita",
    "No unit": "No unit",
}

# The Fingertips API gives a unit's label, not its name: each listed unit's name by its label.
# "£" is left out, as Pholio's "£ per 100,000" has that label too.
UNIT_NAMES_BY_LABEL = {
    "%": "Percent",
    "per 100": "per 100",
    "per 1,000": "per 1,000",
    "per 10,000": "per 10,000",
    "per 100,000": "per 100,000",
    "per 1,000,000": "per 1,000,000",
    "min": "Minutes",
    "Days": "Days",
    "Years": "Years",
    "£ per capita": "£ per capita",
    "No unit": "No unit",
}

UNKNOWN_UNIT_PREFIX = "Unknown unit "

UNIT_OTHER_MAX_LENGTH = 100

UNIT_COLUMNS = ["unit_id", "unit_other"]


def value_type_id(name):
    """The value type for a Fingertips value type name; a name with none stops the export."""
    try:
        return VALUE_TYPE_IDS[VALUE_TYPES[name]]
    except KeyError:
        raise ValueError(f"No value type for the Fingertips value type {name!r}") from None


def unit_values(name):
    """The unit columns' values for a Fingertips unit name; a placeholder stops the export."""
    if name in UNITS:
        return {"unit_id": UNIT_IDS[UNITS[name]], "unit_other": None}
    other = name.strip()
    if not other or other.startswith(UNKNOWN_UNIT_PREFIX) or len(other) > UNIT_OTHER_MAX_LENGTH:
        raise ValueError(f"No unit for the Fingertips unit {name!r}")
    return {"unit_id": UNIT_IDS["Other"], "unit_other": other}


def unit_name_for_label(label):
    """The Fingertips unit name for a Fingertips API unit label; any other label stops the import."""
    try:
        return UNIT_NAMES_BY_LABEL[label]
    except KeyError:
        raise ValueError(f"No Fingertips unit name known for the unit label {label!r}") from None


def _sql_literal(value):
    return "NULL" if value is None else "'" + value.replace("'", "''") + "'"


def value_type_select(value_type_id_expression):
    """The translation as a SELECT column, from an expression holding Pholio's value type id."""
    name = f"(SELECT name FROM value_type WHERE id = {value_type_id_expression})"
    cases = " ".join(
        f"WHEN {_sql_literal(source)} THEN {_sql_literal(value_type_id(source))}"
        for source in VALUE_TYPES
    )
    return f"CASE {name} {cases} END AS value_type_id"


def unit_select(unit_id_expression):
    """The translation as SELECT columns, from an expression holding Pholio's unit id."""
    name = f"(SELECT name FROM unit WHERE id = {unit_id_expression})"
    known = {source: unit_values(source)["unit_id"] for source in UNITS}
    cases = " ".join(
        f"WHEN {_sql_literal(source)} THEN {_sql_literal(unit)}" for source, unit in known.items()
    )
    other = _sql_literal(UNIT_IDS["Other"])
    return (
        f"CASE {name} {cases} ELSE {other} END AS unit_id, "
        f"CASE WHEN {name} IN ({', '.join(_sql_literal(source) for source in UNITS)}) THEN NULL "
        f"ELSE btrim({name}) END AS unit_other"
    )


def unknown_value_types(names):
    """The names among these the translation has no value type for."""
    return sorted(name for name in names if name and name not in VALUE_TYPES)


def unknown_units(names):
    """The names among these the translation has no unit for."""
    unknown = []
    for name in filter(None, names):
        try:
            unit_values(name)
        except ValueError:
            unknown.append(name)
    return sorted(unknown)
